C++: Add dataflow through deep struct indirections

MathiasVP opened the pull request:
a.k.a. implement reverse reads a.k.a. revive a better version of https://github.com/github/codeql/pull/3419.

This PR adds a new dataflow node for each `FieldAddressInstruction` and adds store (read) steps for pushing (popping) field content during stores (load) of deep struct members.

I've tried to make this PR reviewable on a commit-by-commit basis. I'm still hunting down the source of the new inconsistency errors but hopefully, that fix won't require too much restructuring, so I think it's fine to start reviewing this PR as it is.

The ideas in this PR are similar to https://github.com/github/codeql/pull/3419, but instead of adding separate nodes for stores and reads, this PR adds a single dataflow node that serves for both stores and loads.

CPP-difference: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1460/

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:
````diff
@@ -412,16 +365,73 @@ private predicate exactReadStep(Node node1, ArrayContent a, Node node2) {
+ * ```
+ * generates the access path `[z, y, x]`. This predicate ensures that there's a read step from
+ * `x` (the pre update node of `y`) to `y` (the pre update node of `z`), which means that the
+ * shared dataflow library will infer a store step from `z` to `y`.
````
In AST flow it's been convenient that the shared data-flow library adds "reverse read steps" (store steps) automatically because those reverse steps are derived from forward steps that have value on their own and would be there in any case. If it's a hassle to add the forward steps in IR data flow, and if we don't actually need them in their forward form, could we then just extend `storeStep` ourselves instead of relying on reverse reads?

MathiasVP replied:
We could do that, yes. I don't think one is more of a hassle than the other. As it's currently implemented in this PR, we still need the `reverseReadStep` predicate to pop `c` off the access path in `sink(a.b.c)` (here, the pre-update node of `c` is `b`, and so `reverseReadStep(b, _, c)` holds)

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -649,6 +649,36 @@ Node uninitializedNode(LocalVariable v) { none() }
+  exists(FieldNode fieldNode, AddressOperand addressOperand |
+    nodeFrom = fieldNode and
+    not exists(fieldNode.getObjectNode()) and
+    fieldNode.getNextNode*() = getFieldNodeForFieldInstruction(addressOperand.getDef()) and
```
How's the performance of this `*` on kamailio?

MathiasVP replied:
It ran fine on Kamailio. It doesn't stand out in the clause timing report for any of the stages.

jbj reviewed (commented):
This is only a first-pass superficial review.

MathiasVP commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -412,16 +365,73 @@ private predicate exactReadStep(Node node1, ArrayContent a, Node node2) {
   )
 }
 
+pragma[noinline]
+private predicate getField(FieldContent fc, Field field) {
+  exists(Class c |
+    c = field.getDeclaringType() and
+    fc.getAField() = field and
+    fc.hasOffset(c, _, _)
+  )
+}
```
I really only added it to solve an unfortunate join-order. When we write:
```codeql
fieldContent.getAField() = fieldNode.getField()
```
we're allowing `fieldContent` to be all combinations of offsets `n` and classes `c` such that the `n`'th byte of class `c` is the field `fieldNode.getField()` - but we're only interesting in the `fieldContent` where the base class is the declaring class of `fieldNode.getField()`. (i.e., if we write `a.b.c = source()` the `FieldContent` corresponding to `c` should be `(B, offset of c in B)` and not `(A, offset of c in A)`). So we need to write 
```codeql
fieldContent.getAField() = fieldNode.getField() and
fc.hasOffset(fieldNode.getField().getDeclaringType(), _, _)
```
and this tends to create a bad join-order.

But now that you mention it, it did end up looking a lot like the base case of `FieldContent.getAField`. I'll move a version of this predicate into `FieldContent` as a member.

jbj commented:
@rdmarsh2, can you review this PR when you're back from holiday? You reviewed the original PR, #3419.

MathiasVP commented:
6314886e8b7d744d5f3d379f2e3e9d896b02a8b3 fixed some of the problems with multiple post update nodes, but [there are still some left](https://github.com/github/codeql/pull/4350/files?file-filters%5B%5D=.expected#diff-9b2d7a5ac8bf1c94551e0fbbf46acf88R211). These appear to cause problems on `openjdk` just like https://github.com/github/codeql/pull/3419 did. I'll need to figure out how to fix these before we move forward with this PR.

rdmarsh2 commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -202,49 +202,39 @@ private class ArrayContent extends Content, TArrayContent {
-
-private predicate fieldStoreStepChi(Node node1, FieldContent f, PostUpdateNode node2) {
-  exists(StoreInstruction store, ChiInstruction chi |
+/**
+ * A store step from a `StoreInstruction` or `WriteSideEffectInstruction` to the "depest" field
+ * of the destination
+ */
```
Does "deepest" here mean the same thing as "innermost" above? Also, this comment should say something about chi nodes.

MathiasVP replied:
Yes, "deepest" here means "innermost". I'll make that consistent across the comments so that they all say "innermost". Also a good point about the Chi instruction. I will clarify that!

MathiasVP replied:
Fixed in df5ae43cf463e309b3ab735d86149919b3c45751.

rdmarsh2 reviewed (commented):
I'm concerned about `FieldNode` being attached to a `FieldAddressInstruction` rather than a store or load. If the result of a `FieldAddressInstruction` is stored in a variable, there can be multiple `LoadInstructions` that copy from that single store. What happens if we have something like `B *p = &a.b; p->c = 0; p->d = 1; p->e = p->c`?

MathiasVP commented:
> I'm concerned about `FieldNode` being attached to a `FieldAddressInstruction` rather than a store or load. If the result of a `FieldAddressInstruction` is stored in a variable, there can be multiple `LoadInstructions` that copy from that single store. What happens if we have something like `B *p = &a.b; p->c = 0; p->d = 1; p->e = p->c`?

That's a great example of a test we haven't yet added. Since a `FieldNode` is tied to an operand I think your example works as expected.

The assignment `p->e = p->c` generates the following IR:

```
r1(glval<B *>) = VariableAddress[p]       :
r2(B *)        = Load                     : &:r1, m1
r3(glval<int>) = FieldAddress[c]          : r2
r4(int)        = Load                     : &:r3, ~m0
r5(glval<B *>) = VariableAddress[p]       :
r6(B *)        = Load                     : &:r5, m1
r7(glval<int>) = FieldAddress[e]          : r6
m2(int)        = Store                    : &:r7, r4
m3(unknown)    = Chi                      : total:m3, partial:m2
```
In this case, there will be a read step from the definition of `m0` to `FieldNode(r3)`, and from there a `simpleFieldNodeLocalFlowStep` step to the `LoadInstruction` with result `r4`.
Then there will be a simple step from the `LoadInstruction` with result `r4` to the `StoreInstruction` with result `m2`, and from there a store step to `FieldNode(r7)`, and finally there we will have a `simpleFieldNodeLocalFlowStep` to the `ChiInstruction` with result `m3`.

I've added two such test in ddfe0a97531350b4eafab508575dadb68cee661b. The first test is pretty much what you suggested (I hope):
```cpp
void call_sink(B* p) {
  sink(p->e);
}

void multiple_loads_from_a_single_store() {
  A a;
  B *p = &a.b;
  p->c = source();
  p->e = p->c;
  call_sink(p);
}

```
here's the path explanation we get:
```
#   23| call to source
#-----|  -> Store

#   24| Store
#-----|  -> e [e]

#   24| e [e]
#-----|  -> *p [e]

#   16| *p [e]
#-----|  -> e

#   17| e
#-----|  -> e

#   17| e
```
it's a bit of a cheat since the IR infers that the read of `p->c` before the call to `call_sink` reads from the exact same location as was written to on the line above. So we don't really exercise the dataflow library in the way I think you intended this test to do.

The second test I've added is a slight modification:

```cpp
void assign(B* p) {
  p->e = p->c;
}

void multiple_loads_from_a_single_store() {
  A a;
  B *p = &a.b;
  p->c = source();
  assign(p);
  call_sink(p);
}
```

and here we get the following path explanation:
```
#   23| call to source
#-----|  -> Store

#   23| Store
#-----|  -> c [c]

#   23| c [c]
#-----|  -> assign output argument [e]

#   24| assign output argument [e]
#-----|  -> *p [e]

#   16| *p [e]
#-----|  -> e

#   17| e
#-----|  -> e

#   17| e
```

MathiasVP commented:
Locally I've convinced myself that 032fcdc54fab25bd02d2a4d049bec7e3824972b2 fixes the performance problems on `openjdk`. I've started a CPP-difference that should hopefully confirm this: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1495/

rdmarsh2 commented:
Code looks good to me. I'll wait to merge until the CPP-Differences is done (or leave it for @jbj tomorrow)

MathiasVP commented:
Tiny update: The problematic function on `openjdk` that blows up in IR dataflow:

```cpp
  AD.generateMatchLists();
  AD.identify_unique_operands();
  AD.identify_cisc_spill_instructions();
  AD.identify_short_branches();
  // Make sure every file starts with a copyright:
  AD.addSunCopyright(legal_text, legal_sz, AD._HPP_file._fp);           // .hpp
  AD.addSunCopyright(legal_text, legal_sz, AD._CPP_file._fp);           // .cpp
  AD.addSunCopyright(legal_text, legal_sz, AD._CPP_CLONE_file._fp);     // .cpp
  AD.addSunCopyright(legal_text, legal_sz, AD._CPP_EXPAND_file._fp);    // .cpp
  AD.addSunCopyright(legal_text, legal_sz, AD._CPP_FORMAT_file._fp);    // .cpp
  AD.addSunCopyright(legal_text, legal_sz, AD._CPP_GEN_file._fp);       // .cpp
  AD.addSunCopyright(legal_text, legal_sz, AD._CPP_MISC_file._fp);      // .cpp
  AD.addSunCopyright(legal_text, legal_sz, AD._CPP_PEEPHOLE_file._fp);  // .cpp
  AD.addSunCopyright(legal_text, legal_sz, AD._CPP_PIPELINE_file._fp);  // .cpp
  AD.addSunCopyright(legal_text, legal_sz, AD._VM_file._fp);            // .hpp
  AD.addSunCopyright(legal_text, legal_sz, AD._DFA_file._fp);           // .cpp
  // Add include guards for all .hpp files
  AD.addIncludeGuardStart(AD._HPP_file, "GENERATED_ADFILES_AD_HPP");        // .hpp
  AD.addIncludeGuardStart(AD._VM_file, "GENERATED_ADFILES_ADGLOBALS_HPP");  // .hpp
  // Add includes
  AD.addInclude(AD._CPP_file, "precompiled.hpp");
  AD.addInclude(AD._CPP_file, "adfiles", get_basename(AD._VM_file._name));
  AD.addInclude(AD._CPP_file, "adfiles", get_basename(AD._HPP_file._name));
  AD.addInclude(AD._CPP_file, "memory/allocation.inline.hpp");
  ...
  AD.close_files(0);               // Close all input/output files
```

_also_ blows up (but not as bad) in the AST version of the dataflow library. With the IR dataflow the big step that ends at `AD.close_files(0);` has 240 predecessors, and with the AST dataflow the same node has 100 predecessors. I think this means that there's not something fundamentally wrong with the IR version (or rather: the same problem is also present in the AST version), but the problem is amplified because the IR dataflow has "smaller" small steps than the AST (since we do `instruction -> operand -> instruction` steps). Making the source node of read steps break the big step relation (i.e., making these nodes `CastNodes`) reduces the number of predecessors for this node from 240 to 154, and making all non-conflated Chi instructions `CastNodes` reduces the number to 25.

I was hoping that this would fix the problem, and indeed it looks like the local flow steps are fine now. However, now there's a slow tuple explosion (i.e., a long sequence of iterations that slowly blows up) in `flowFwd0` in this disjunction:

```codeql
    exists(AccessPath argAp0 |
      flowFwdOutFromArg(call, node, argAp0, apf, ap, config) and
      flowFwdIsEntered(call, cc, argAp, argAp0, config)
    )
```

MathiasVP commented:
Superseded by https://github.com/github/codeql/pull/4784.
